pkgs/pkgs-lib/formats.nix

```diff
@@ -147,4 +156,202 @@ rec {
+    the Elixir section of the Nixos documentation.
+
+    TODO: special Elixir values doesn't show up nicely in the documentation
+
+    [Config]: <https://hexdocs.pm/elixir/Config.html>
+    [Keyword list]: <https://hexdocs.pm/elixir/Keyword.html>
+    [Map]: <https://hexdocs.pm/elixir/Map.html>
+    [String]: <https://hexdocs.pm/elixir/String.html>
+    [Atom]: <https://hexdocs.pm/elixir/Atom.html>
+    [List]: <https://hexdocs.pm/elixir/List.html>
+    [Tuple]: <https://hexdocs.pm/elixir/Tuple.html>
+  */
+  elixirConf = { elixir ? pkgs.elixir }:
+    with lib; let
+      toElixir = value: with builtins;
+        if value == null then "nil" else
+        if value == true then "true" else
+        if value == false then "false" else
+        if isInt value || isFloat value then toString value else
+        if isString value then string value else
+        if isAttrs value then attrs value else
+        if isList value then list value else
+        abort "formats.elixirConf: should never happen (value = ${value})";
+
+      escapeElixir = escape [ "\\" "#" "\"" ];
+      string = value: "\"${escapeElixir value}\"";
+
+      attrs = set:
+        if set ? _elixirType then specialType set
+        else
+          let
+            toKeyword = name: value: "${name}: ${toElixir value}";
+            keywordList = concatStringsSep ", " (mapAttrsToList toKeyword set);
+          in
+          "[" + keywordList + "]";
+
+      listContent = values: concatStringsSep ", " (map toElixir values);
+
+      list = values: "[" + (listContent values) + "]";
+
+      specialType = { value, _elixirType }:
+        if _elixirType == "raw" then value else
+        if _elixirType == "atom" then value else
+        if _elixirType == "map" then elixirMap value else
+        if _elixirType == "tuple" then tuple value else
+        abort "formats.elixirConf: should never happen (_elixirType = ${_elixirType})";
+
+      elixirMap = set:
+        let
+          toEntry = name: value: "${toElixir name} => ${toElixir value}";
+          entries = concatStringsSep ", " (mapAttrsToList toEntry set);
+        in
+        "%{${entries}}";
+
+      tuple = values: "{${listContent values}}";
+
+      toConf = values:
+        let
+          keyConfig = rootKey: key: value:
+            "config ${rootKey}, ${key}, ${toElixir value}";
+          keyConfigs = rootKey: values: mapAttrsToList (keyConfig rootKey) values;
+          rootConfigs = flatten (mapAttrsToList keyConfigs values);
+        in
+        ''
+          import Config
+
+          ${concatStringsSep "\n" rootConfigs}
+        '';
+    in
+    {
+      type = with lib.types; let
+        valueType = nullOr
+          (oneOf [
+            bool
+            int
+            float
+            str
+            (attrsOf valueType)
+            (listOf valueType)
+          ]) // {
+          description = "Elixir value";
+        };
+      in
+      attrsOf (attrsOf (valueType));
```

Comment:
Yes, the first key is the name (or rather atom) of the application to configure, the second is the configuration key, and then you can have whatever configuration value you want.

Reply:
@minijackson I'm stumbling over a case where I think this isn't the case. I want to generate this config:

```exs
import Config

config :logger, level: :info
```

My current code is `(formats.elixirConf { }).generate "test" { ":logger".level = (pkgs.formats.elixirConf { }).lib.mkAtom ":debug"; }`, which results in

```exs
import Config

config :logger, level, :debug
```

(Note the missing colon after `level`.)

I have absolutely no clue about elixir, but it looks to me like there are two implementations of the `config` function. One that takes three arguments and this one: https://hexdocs.pm/elixir/Config.html#config/2

Any idea how I could achieve my goal using the existing implementation? Or what would need to be changed?

Reply:
@erictapen from a quick look at the [implementation](https://github.com/elixir-lang/elixir/blob/v1.13.4/lib/elixir/lib/config.ex#L123-L165=), it seems like `config :logger, level: :debug` should be equivalent to `config :logger, :level, :debug`. So you should be able to get what you want using this:

```nix
{ ":logger".":level" = (pkgs.formats.elixirConf { }).lib.mkAtom ":debug"; }
```

I was worried for a second that there might be cases where the second argument to `config/2` is not a keyword list, but it seems like it is [always the case](https://github.com/elixir-lang/elixir/blob/v1.13.4/lib/elixir/lib/config.ex#L125). Although I agreed this might not be intuitive for end users not familiar with Elixir. I'm not sure what we should do about this, though, maybe a mention of this case in the documentation?

Reply:
Ah cool, yeah for me as an outsider these colons look absolutely random :D
I think the best implementation would throw an eval error, notifying people about the alternative syntax, but I don't see an easy solution to implement that currently.
This already helps in my specific use case, thanks!